File: Data Structures & Algorithms/combination-target-sum-ii/submission-5.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        n = len(candidates)
        res = []
        candidates.sort()
        def util(j, t, r):
            if t < 0:
                return
            if t == 0:
                res.append(r[:])
                return
            for i in range(j, n):
                if candidates[i] > t:
                    break
                if (j < i and candidates[i] == candidates[i-1]):
                    continue
                r.append(candidates[i])
                util(i+1, t-candidates[i], r)
                r.pop()

        util(0, target, [])
        return res

File: Data Structures & Algorithms/combination-target-sum/submission-5.cpp
class Solution {
public:
    vector<vector<int> > combinationSum(vector<int> &candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int> > res;
        vector<int> combination;
        combinationSum(candidates, target, res, combination, 0);
        return res;
    }
private:
    void combinationSum(vector<int> &candidates, int target, vector<vector<int> > &res, vector<int> &combination, int i) {
        if (target < 0) return;
        if (!target) {
            res.push_back(combination);
            return;
        }
        for (; i != candidates.size() && target >= candidates[i]; i++) {
            combination.push_back(candidates[i]);
            combinationSum(candidates, target - candidates[i], res, combination, i);
            combination.pop_back();
        }
    }
};